### PR TITLE
Remove unnecessary `TrainingEpochLoop` return

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

We were returning the `training_epoch_end` input as the output of `TrainingEpochLoop`, I believe this is legacy and no longer necessary. Let's see what CI thinks

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [n/a] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: tests/trainer/test_dataloaders.py [hunk 1]
```diff
@@ -899,6 +899,24 @@ def gen(self):
...
+
+        def gen(self):
+            yield from []
+
+    model = BoringModel()
+    train_dataloader = DataLoader(RandomDataset(), batch_size=2)
+    trainer = Trainer(default_root_dir=os.getcwd(), max_epochs=2)
+    trainer.fit(model, train_dataloader=train_dataloader)
+    assert trainer.global_step == 0
+    assert trainer.current_epoch == 1
```
Comment (carmocca) [inline tests/trainer/test_dataloaders.py:917]: Not sure if this value is correct, two arguments:

- No batch was run so current epch should be `0`
- An epoch composed of 0 batches was successfully run, the current epoch should be `1`

`1` is the value in master

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 2]
```diff
@@ -193,32 +193,29 @@ def on_run_end(self) -> List[List[STEP_OUTPUT]]:
...
         """
-        if self.batch_progress.current.ready == 0:
-            # dataloader/iterator did not produce a batch
-            return
-
         # inform logger the batch loop has finished
         self.trainer.logger_connector.epoch_end_reached()
 
-        # prepare epoch output
-        processed_outputs = self._prepare_outputs(self._epoch_output, batch_mode=False)
```
Comment (carmocca) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:204]: This wasn't a memory leak because `self._epoch_output` would be composed of empty lists if `training_epoch_end` is not overridden

File: tests/trainer/test_dataloaders.py (hunk 1 above)
Comment (awaelchli) [reply tests/trainer/test_dataloaders.py:917]: hey, when I fixed this specific edge case I added a test for it: 

`test_iterable_dataset_stop_iteration_at_epoch_beginning`

it may be simpler to adapt this one. whether the iterator is empty from the beginning of the training or the beginning of an epoch should not matter

Comment (carmocca) [reply tests/trainer/test_dataloaders.py:897]: Merged

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 3]
```diff
@@ -193,32 +193,29 @@ def on_run_end(self) -> List[List[STEP_OUTPUT]]:
...
-                    "training_epoch_end expects a return of None. "
-                    "HINT: remove the return statement in training_epoch_end"
-                )
+        if is_overridden("training_epoch_end", model) and self._epoch_output:
+            processed_outputs = self._prepare_outputs(self._epoch_output, batch_mode=False)
+            # check that the dataloader/iterator produced a batch
+            if processed_outputs:
+                # run training_epoch_end
+                # refresh the result for custom logging at the epoch level
+                model._current_fx_name = "training_epoch_end"
```
Comment (tchaton) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:207]: Can this be replaced by call_hook ?

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:207]: yes, it will

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 4]
```diff
@@ -193,32 +193,29 @@ def on_run_end(self) -> List[List[STEP_OUTPUT]]:
         Raises:
             MisconfigurationException: ``train_epoch_end`` does not return ``None``
         """
-        if self.batch_progress.current.ready == 0:
```
Comment (awaelchli) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:196]: if I remove this check in the commit of PR #7871 then the test
`test_iterable_dataset_stop_iteration_at_epoch_beginning`
fails, however here it does not.

just an observation.

**Review by awaelchli — APPROVED** [review]
looks like we are covering the edge case, hope we don't miss another one 🙈

**Review by Borda — APPROVED** [review]
